### PR TITLE
Fix starvation caused by resource_overcommit

### PR DESCRIPTION
**Pull request by cberner**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java [hunk 1]
```diff
@@ -143,21 +143,17 @@ public synchronized void process(Iterable<QueryExecution> queries)
...
-                else {
-                    DataSize maxMemory = succinctDataSize(queryMemoryLimit, BYTE);
-                    query.fail(exceededGlobalLimit(maxMemory));
-                    queryKilled = true;
-                }
+            if (resourceOvercommit(query.getSession()) && outOfMemory) {
+                // If a query has requested resource overcommit, only kill it if the cluster has run out of memory
+                DataSize memory = succinctDataSize(bytes, BYTE);
+                query.fail(new PrestoException(CLUSTER_OUT_OF_MEMORY,
+                        format("The cluster is out of memory and you set %s=true, so your query was killed. It was using %s of memory.", RESOURCE_OVERCOMMIT, memory)));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java]: I don't like error messages with the word `you` in them

File: presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java [hunk 2]
```diff
@@ -143,21 +143,17 @@ public synchronized void process(Iterable<QueryExecution> queries)
...
-                                format("The cluster is out of memory, you set %s=true, and your query is using %s of memory, so it was killed.", RESOURCE_OVERCOMMIT, memory)));
-                        queryKilled = true;
-                    }
-                }
-                else {
-                    DataSize maxMemory = succinctDataSize(queryMemoryLimit, BYTE);
-                    query.fail(exceededGlobalLimit(maxMemory));
-                    queryKilled = true;
-                }
+            if (resourceOvercommit(query.getSession()) && outOfMemory) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java]: So if you use resource over commit, than your query can be killed any time the system is low on memory, even if your query isn't actually using much memory?

**Comment by dain** [comment]
Assuming this is how you want resource over commit to work, looks good to me.

File: presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java (hunk 2 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java]: Yes, the original code guaranteed that you got your fair share, but it is buggy and implementing that check correctly would be complicated

**Pull request merged**